src/main/java/org/reaktivity/nukleus/http2/internal/routable/stream/SourceInputStreamFactory.java

```diff
@@ -655,24 +651,45 @@ void error(Http2ErrorCode errorCode)
             replyTarget.doEnd(sourceOutputEstId);
         }
 
-        private void doPromisedRequest(int http2StreamId, Map<String, String> headersMap)
+        private IntSupplier promisedStreamIds()
         {
+            return () ->
+            {
+                lastPromisedStreamId += 2;
+                return lastPromisedStreamId;
+            };
+        };
+
+        private void doPromisedRequest(int http2StreamId, ListFW<HttpHeaderFW> headers)
+        {
+            Map<String, String> headersMap = new HashMap<>();
+            headers.forEach(
+                    httpHeader -> headersMap.put(httpHeader.name().asString(), httpHeader.value().asString()));
             Optional<Route> optional = resolveTarget(sourceRef, headersMap);
             Route route = optional.get();
             Target newTarget = route.target();
             long targetRef = route.targetRef();
             long targetId = supplyStreamId.getAsLong();
             newTarget.doHttpBegin(targetId, targetRef, targetId,
-                    hs -> headersMap.forEach((k, v) -> hs.item(i -> i.representation((byte) 0).name(k).value(v))));
+                    hs -> headers.forEach(h -> hs.item(builder -> mapHeader(h, builder))));
             Correlation correlation = new Correlation(correlationId, sourceOutputEstId, this::doPromisedRequest,
-                    http2StreamId, promisedStreamIds, source.routableName(), OUTPUT_ESTABLISHED);
+                    http2StreamId, encodeContext, promisedStreamIds(), source.routableName(), OUTPUT_ESTABLISHED);
 
             correlateNew.accept(targetId, correlation);
             newTarget.addThrottle(targetId, this::handleThrottle);
 
             Http2Stream http2Stream = new Http2Stream(this, http2StreamId, targetId, State.HALF_CLOSED_REMOTE);
             http2Streams.put(http2StreamId, http2Stream);
         }
+
+        private void mapHeader(HttpHeaderFW h, HttpHeaderFW.Builder builder)
+        {
+            StringFW name = h.name();
+            StringFW value = h.value();
+            builder.representation((byte) 0)
+                   .name(name.buffer(), name.offset()+1, name.sizeof()-1)
+                   .value(value.buffer(), value.offset()+1, value.sizeof()-1);
```

Comment:
We have added support for DirectBuffer, we can also add StringFW. Filed
https://github.com/reaktivity/nukleus-maven-plugin/issues/19